### PR TITLE
Script to pull repos based on config.

### PR DESCRIPTION
**Pull request by aleffert**
JIRA: https://openedx.atlassian.net/browse/MA-1514

### CODE REVIEW

**Comment by aleffert** [comment]
@benpatterson @nasthagiri please review

File: mobile_app/README.md [hunk 1]
```diff
@@ -29,3 +29,19 @@ separate build machine.
 The expectation is that the branch name will be some unique identifier like the
 jenkins job number and the date.
 	
+
+checkout_build_repos
+--------------------
+
+Checks out out a code and config repository and sets them up for building, by creating a properties file to point the code at the config.
```
Comment (nasthagiri) [inline mobile_app/README.md]: "Checks *out out* a code and config repository" -> "Checks out code and config repositories"

File: mobile_app/checkout_build_repos.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+def _write_properties(
+        code_repo_properties_path,
+        relative_config_path,
+        code_repo_path,
+        config_repo_path
+        ):
+    """
+    Writes an edx.properties file to the code repo
+    that points that points at the proper location in the config
```
Comment (nasthagiri) [inline mobile_app/checkout_build_repos.py]: s/that points that points/that points

File: mobile_app/path_constants.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+"""
+Path constants used by the mobile build scripts
+"""
+CONFIG_FILE = "CONFIGURATION"
+
+CODE_CHECKOUT = "code.git"
```
Comment (nasthagiri) [inline mobile_app/path_constants.py]: Suggestion to rename these to CHECKOUT_DIRECTORY_NAME so it's clear it's the name of the local directory to where they are checked out.

File: mobile_app/checkout_build_repos.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    CODE_CHECKOUT,
+    CONFIG_CHECKOUT,
+    CONFIG_FILE,
+    PROPERTIES_FILE
+)
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+def _write_properties(
```
Comment (nasthagiri) [inline mobile_app/checkout_build_repos.py]: Nit: Per clean code, move the internal helper methods below the top-level methods.

**Comment by benpatterson** [comment]
I realize this is a somewhat fundamental question, but did you consider handling this through ansible? That has a git module and can do a lot of what's described in here.

Jenkins can likewise checkout multiple repos, but which repos are involved is less dynamic.

**Comment by aleffert** [comment]
@benpatterson oh that makes sense. I don't really know too much about this area. I know we went through the tasks with you at some point, but maybe it's worth going over a little more before going too much farther.

Note that it doesn't make sense to do this on jenkins since this part is executing on travis.

**Comment by benpatterson** [comment]
Ah yes I forgot about the travis part, @aleffert. 

Ansible manages some of this for the analytics team in their jobs (spins up and builds out AWS EMR instances). Not sure how much ansible you've slung around, but [here](http://docs.ansible.com/ansible/git_module.html) is their documentation, and [here](https://github.com/edx/configuration/blob/master/playbooks/roles/jenkins_worker/tasks/python_platform_worker.yml#L8-L14) is a basic example of it, though you could easily replace some of those with variables, and likewise loop through a list of them.

**Comment by aleffert** [comment]
Talked to @benpatterson offline. Decided to proceed with python for now

File: mobile_app/checkout_build_repos.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+    # checkout repos
+    repos = [
+        (code_repo_path, code_repo_url, code_revision),
+        (config_repo_path, config_repo_url, config_revision)
+    ]
+    for (repo_path, url, revision) in repos:
+
+        logger.info("cloning repo %s to %s", url, repo_path)
+        repo = git.Repo.clone_from(url, to_path=repo_path)
```
Comment (benpatterson) [inline mobile_app/checkout_build_repos.py]: Do you want to retry if there's a timeout or some other kind of error?

File: mobile_app/checkout_build_repos.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        origin_branch = repo.remotes.origin.refs[revision]
+        # need to create a local tracking branch of the remote
+        # so we can check it out properly
+        repo.create_head(revision)
```
Comment (benpatterson) [inline mobile_app/checkout_build_repos.py]: I am not familiar with this particular method. What happens if this revision already exists on remote?

Comment (benpatterson) [inline mobile_app/checkout_build_repos.py]: ...and likewise, if it does not exist?

**Comment by benpatterson** [comment]
I have some questions about some scenarios. The code itself LGTM.

File: mobile_app/checkout_build_repos.py (hunk 6 above)
Comment (aleffert) [inline mobile_app/checkout_build_repos.py]: Remote refs have a different namespace than local refs. If the remote ref does not exist then line 116 will throw a branch not found exception

If the local branch does already exist then ``create_head`` will throw an exception, both of those cases are expected failure cases

File: mobile_app/checkout_build_repos.py (hunk 5 above)
Comment (aleffert) [inline mobile_app/checkout_build_repos.py]: I would expect those to be rare, and I think it's fine if the user kicks off the job again. If we start doing nightlies and this is a problem, we can revisit it, but I'd guess travis can't download a repo from github is probably going to be more than a momentary glitch

**Comment by aleffert** [comment]
@benpatterson @nasthagiri comments addressed

**Comment by benpatterson** [comment]
:+1:

**Comment by aleffert** [comment]
psst @nasthagiri. Anything else before I merge?

**Comment by nasthagiri** [comment]
:+1: looks good

**Pull request merged**